File: src/android/app/src/main/java/org/citra/citra_emu/features/cheats/model/CheatsViewModel.java
package org.citra.citra_emu.features.cheats.model;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class CheatsViewModel extends ViewModel {
    private int mSelectedCheatPosition = -1;
    private final MutableLiveData<Cheat> mSelectedCheat = new MutableLiveData<>(null);
    private final MutableLiveData<Boolean> mIsAdding = new MutableLiveData<>(false);
    private final MutableLiveData<Boolean> mIsEditing = new MutableLiveData<>(false);

    private final MutableLiveData<Integer> mCheatAddedEvent = new MutableLiveData<>(null);
    private final MutableLiveData<Integer> mCheatChangedEvent = new MutableLiveData<>(null);
    private final MutableLiveData<Integer> mCheatDeletedEvent = new MutableLiveData<>(null);
    private final MutableLiveData<Boolean> mOpenDetailsViewEvent = new MutableLiveData<>(false);

    private Cheat[] mCheats;
    private boolean mCheatsNeedSaving = false;

    public void load() {
        mCheats = CheatEngine.getCheats();

        for (int i = 0; i < mCheats.length; i++) {
            int position = i;
            mCheats[i].setEnabledChangedCallback(() -> {
                mCheatsNeedSaving = true;
                notifyCheatUpdated(position);
            });
        }
    }

    public void saveIfNeeded() {
        if (mCheatsNeedSaving) {
            CheatEngine.saveCheatFile();
            mCheatsNeedSaving = false;
        }
    }

    public Cheat[] getCheats() {
        return mCheats;
    }

    public LiveData<Cheat> getSelectedCheat() {
        return mSelectedCheat;
    }

    public void setSelectedCheat(Cheat cheat, int position) {
        if (mIsEditing.getValue()) {
            setIsEditing(false);
        }

        mSelectedCheat.setValue(cheat);
        mSelectedCheatPosition = position;
    }

    public LiveData<Boolean> getIsAdding() {
        return mIsAdding;
    }

    public LiveData<Boolean> getIsEditing() {
        return mIsEditing;
    }

    public void setIsEditing(boolean isEditing) {
        mIsEditing.setValue(isEditing);

        if (mIsAdding.getValue() && !isEditing) {
            mIsAdding.setValue(false);
            setSelectedCheat(null, -1);
        }
    }

    /**
     * When a cheat is added, the integer stored in the returned LiveData
     * changes to the position of that cheat, then changes back to null.
     */
    public LiveData<Integer> getCheatAddedEvent() {
        return mCheatAddedEvent;
    }

    private void notifyCheatAdded(int position) {
        mCheatAddedEvent.setValue(position);
        mCheatAddedEvent.setValue(null);
    }

    public void startAddingCheat() {
        mSelectedCheat.setValue(null);
        mSelectedCheatPosition = -1;

        mIsAdding.setValue(true);
        mIsEditing.setValue(true);
    }

    public void finishAddingCheat(Cheat cheat) {
        if (!mIsAdding.getValue()) {
            throw new IllegalStateException();
        }

        mIsAdding.setValue(false);
        mIsEditing.setValue(false);

        int position = mCheats.length;

        CheatEngine.addCheat(cheat);

        mCheatsNeedSaving = true;
        load();

        notifyCheatAdded(position);
        setSelectedCheat(mCheats[position], position);
    }

    /**
     * When a cheat is edited, the integer stored in the returned LiveData
     * changes to the position of that cheat, then changes back to null.
     */
    public LiveData<Integer> getCheatUpdatedEvent() {
        return mCheatChangedEvent;
    }

    /**
     * Notifies that an edit has been made to the contents of the cheat at the given position.
     */
    private void notifyCheatUpdated(int position) {
        mCheatChangedEvent.setValue(position);
        mCheatChangedEvent.setValue(null);
    }

    public void updateSelectedCheat(Cheat newCheat) {
        CheatEngine.updateCheat(mSelectedCheatPosition, newCheat);

        mCheatsNeedSaving = true;
        load();

        notifyCheatUpdated(mSelectedCheatPosition);
        setSelectedCheat(mCheats[mSelectedCheatPosition], mSelectedCheatPosition);
    }

    /**
     * When a cheat is deleted, the integer stored in the returned LiveData
     * changes to the position of that cheat, then changes back to null.
     */
    public LiveData<Integer> getCheatDeletedEvent() {
        return mCheatDeletedEvent;
    }

    /**
     * Notifies that the cheat at the given position has been deleted.
     */
    private void notifyCheatDeleted(int position) {
        mCheatDeletedEvent.setValue(position);
        mCheatDeletedEvent.setValue(null);
    }

    public void deleteSelectedCheat() {
        int position = mSelectedCheatPosition;

        setSelectedCheat(null, -1);

        CheatEngine.removeCheat(position);

        mCheatsNeedSaving = true;
        load();

        notifyCheatDeleted(position);
    }

    public LiveData<Boolean> getOpenDetailsViewEvent() {
        return mOpenDetailsViewEvent;
    }

    public void openDetailsView() {
        mOpenDetailsViewEvent.setValue(true);
        mOpenDetailsViewEvent.setValue(false);
    }
}

File: src/android/app/src/main/java/org/citra/citra_emu/features/cheats/model/CheatsViewModel.kt
// Copyright 2023 Citra Emulator Project
// Licensed under GPLv2 or any later version
// Refer to the license.txt file included.

package org.citra.citra_emu.features.cheats.model

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

class CheatsViewModel : ViewModel() {
    val selectedCheat get() = _selectedCheat.asStateFlow()
    private val _selectedCheat = MutableStateFlow<Cheat?>(null)

    val isAdding get() = _isAdding.asStateFlow()
    private val _isAdding = MutableStateFlow(false)

    val isEditing get() = _isEditing.asStateFlow()
    private val _isEditing = MutableStateFlow(false)

    /**
     * When a cheat is added, the integer stored in the returned StateFlow
     * changes to the position of that cheat, then changes back to null.
     */
    val cheatAddedEvent get() = _cheatAddedEvent.asStateFlow()
    private val _cheatAddedEvent = MutableStateFlow<Int?>(null)

    val cheatChangedEvent get() = _cheatChangedEvent.asStateFlow()
    private val _cheatChangedEvent = MutableStateFlow<Int?>(null)

    /**
     * When a cheat is deleted, the integer stored in the returned StateFlow
     * changes to the position of that cheat, then changes back to null.
     */
    val cheatDeletedEvent get() = _cheatDeletedEvent.asStateFlow()
    private val _cheatDeletedEvent = MutableStateFlow<Int?>(null)

    val openDetailsViewEvent get() = _openDetailsViewEvent.asStateFlow()
    private val _openDetailsViewEvent = MutableStateFlow(false)

    val closeDetailsViewEvent get() = _closeDetailsViewEvent.asStateFlow()
    private val _closeDetailsViewEvent = MutableStateFlow(false)

    val listViewFocusChange get() = _listViewFocusChange.asStateFlow()
    private val _listViewFocusChange = MutableStateFlow(false)

    val detailsViewFocusChange get() = _detailsViewFocusChange.asStateFlow()
    private val _detailsViewFocusChange = MutableStateFlow(false)

    private var titleId: Long = 0
    lateinit var cheats: Array<Cheat>
    private var cheatsNeedSaving = false
    private var selectedCheatPosition = -1

    fun initialize(titleId_: Long) {
        titleId = titleId_;
        load()
    }

    private fun load() {
        CheatEngine.loadCheatFile(titleId)
        cheats = CheatEngine.getCheats()
        for (i in cheats.indices) {
            cheats[i].setEnabledChangedCallback {
                cheatsNeedSaving = true
                notifyCheatUpdated(i)
            }
        }
    }

    fun saveIfNeeded() {
        if (cheatsNeedSaving) {
            CheatEngine.saveCheatFile(titleId)
            cheatsNeedSaving = false
        }
    }

    fun setSelectedCheat(cheat: Cheat?, position: Int) {
        if (isEditing.value) {
            setIsEditing(false)
        }
        _selectedCheat.value = cheat
        selectedCheatPosition = position
    }

    fun setIsEditing(value: Boolean) {
        _isEditing.value = value
        if (isAdding.value && !value) {
            _isAdding.value = false
            setSelectedCheat(null, -1)
        }
    }

    private fun notifyCheatAdded(position: Int) {
        _cheatAddedEvent.value = position
        _cheatAddedEvent.value = null
    }

    fun startAddingCheat() {
        _selectedCheat.value = null
        selectedCheatPosition = -1
        _isAdding.value = true
        _isEditing.value = true
    }

    fun finishAddingCheat(cheat: Cheat?) {
        check(isAdding.value)
        _isAdding.value = false
        _isEditing.value = false
        val position = cheats.size
        CheatEngine.addCheat(cheat)
        cheatsNeedSaving = true
        load()
        notifyCheatAdded(position)
        setSelectedCheat(cheats[position], position)
    }

    /**
     * Notifies that an edit has been made to the contents of the cheat at the given position.
     */
    private fun notifyCheatUpdated(position: Int) {
        _cheatChangedEvent.value = position
        _cheatChangedEvent.value = null
    }

    fun updateSelectedCheat(newCheat: Cheat?) {
        CheatEngine.updateCheat(selectedCheatPosition, newCheat)
        cheatsNeedSaving = true
        load()
        notifyCheatUpdated(selectedCheatPosition)
        setSelectedCheat(cheats[selectedCheatPosition], selectedCheatPosition)
    }

    /**
     * Notifies that the cheat at the given position has been deleted.
     */
    private fun notifyCheatDeleted(position: Int) {
        _cheatDeletedEvent.value = position
        _cheatDeletedEvent.value = null
    }

    fun deleteSelectedCheat() {
        val position = selectedCheatPosition
        setSelectedCheat(null, -1)
        CheatEngine.removeCheat(position)
        cheatsNeedSaving = true
        load()
        notifyCheatDeleted(position)
    }

    fun openDetailsView() {
        _openDetailsViewEvent.value = true
        _openDetailsViewEvent.value = false
    }

    fun closeDetailsView() {
        _closeDetailsViewEvent.value = true
        _closeDetailsViewEvent.value = false
    }

    fun onListViewFocusChanged(changed: Boolean) {
        _listViewFocusChange.value = changed
        _listViewFocusChange.value = false
    }

    fun onDetailsViewFocusChanged(changed: Boolean) {
        _detailsViewFocusChange.value = changed
        _detailsViewFocusChange.value = false
    }
}
